aptos-move/framework/aptos-framework/sources/AptosGovernance.move

```diff
@@ -0,0 +1,360 @@
+        // has voted. This doesn't take into subsequent inflation/deflation (rewards are issued every epoch and gas fees
+        // are burnt after every transaction), but inflation/delation is very unlikely to have a major impact on total
+        // supply during the voting period.
+        let total_voting_token_supply = Coin::supply<TestCoin>();
+        let early_resolution_vote_threshold = Option::none<u128>();
+        if (Option::is_some(&total_voting_token_supply)) {
+            let total_supply = *Option::borrow(&total_voting_token_supply);
+            // 50% + 1 to avoid rounding errors.
+            early_resolution_vote_threshold = Option::some(total_supply / 2 + 1);
+        };
+
+        let proposal_id = Voting::create_proposal(
+            @AptosFramework,
+            GovernanceProposal::create_proposal(),
+            Option::some(execution_hash),
+            governance_config.min_voting_threshold,
+            proposal_expiration,
+            early_resolution_vote_threshold,
+        );
+
+        let events = borrow_global_mut<GovernanceEvents>(@AptosFramework);
+        Event::emit_event<CreateProposalEvent>(
+            &mut events.create_proposal_events,
+            CreateProposalEvent {
+                proposal_id,
+                proposer: proposer_address,
+                stake_pool,
+                execution_hash,
+            },
+        );
+    }
+
+    /// Vote on proposal with `proposal_id` and voting power from `stake_pool`.
+    public(script) fun vote(
+        voter: &signer,
+        stake_pool: address,
+        proposal_id: u64,
+        should_pass: bool,
+    ) acquires GovernanceEvents, VotingRecords {
+        let voter_address = Signer::address_of(voter);
+        assert!(Stake::is_delegated_voter(stake_pool, voter_address), Errors::invalid_argument(ENOT_DELEGATED_VOTER));
+
+        // The voter's stake needs to be locked up at least as long as the proposal's expiration.
+        let proposal_expiration = Voting::get_proposal_expiration_secs<GovernanceProposal>(@AptosFramework, proposal_id);
+        assert!(
+            Stake::get_lockup_secs(stake_pool) >= proposal_expiration,
+            Errors::invalid_argument(EINSUFFICIENT_STAKE_LOCKUP),
+        );
+
+        // Ensure the voter doesn't double vote.
+        let voting_records = borrow_global_mut<VotingRecords>(@AptosFramework);
+        let record_key = RecordKey {
+            stake_pool,
+            proposal_id,
+        };
+        assert!(
+            !Table::contains(&voting_records.votes, record_key),
+            Errors::invalid_argument(EALREADY_VOTED));
+        Table::add(&mut voting_records.votes, record_key, true);
+
+        // Voting power does not include pending_active or pending_inactive balances.
+        // In general, the stake pool should not have pending_inactive balance if it still has lockup (requried to vote)
+        // And if pending_active will be added to active in the next epoch.
```

Comment:
this is kinda interesting...  what if I have more stake after an epoch change... maybe we should allow "increase in vote stake" but not changing the vote?

Reply:
I think it's a slippery slope, users should only vote once for simplicity. They should consolidate and stake all the coins they have before voting.